theme-kit/router.ts

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/naming-convention */
+
+import type { Culture } from '@prezly/sdk';
+import { notFound } from 'next/navigation';
+import type { ReactElement, ReactNode } from 'react';
+import UrlPattern from 'url-pattern';
+
+import { assertServerEnv } from '@/utils';
+
+import { api } from './api';
+
+export function route<
+    Pattern extends `/${string}`,
+    Match extends ExtractPathParams<Pattern>,
+    Props = Match,
+>(pattern: Pattern, loader: Loader<Match, Props>) {
+    const urlPattern = buildUrlPattern(pattern);
+
+    return {
+        async match(path: string, getDefaultLocale, isEnabledLocale) {
+            const matched = urlPattern.match(path);
+
+            if (!matched) {
+                return undefined;
+            }
+
+            if (matched.locale && !(await isEnabledLocale(matched.locale))) {
+                return undefined;
+            }
+
+            if (!matched.locale) {
+                matched.locale = await getDefaultLocale();
+            }
+
+            const { match: pageMatch, resolveLocale, Layout, default: Page } = await loader();
+
+            const pageMatched = pageMatch ? await pageMatch(matched) : matched;
+
+            if (!pageMatched) {
+                return undefined;
+            }
+
+            if (resolveLocale) {
+                return {
+                    match: {
+                        ...pageMatched,
+                        locale: await resolveLocale(pageMatched),
+                    },
+                    Layout,
+                    Page,
+                };
+            }
+
+            return {
+                match: { ...pageMatched, locale: matched.locale ?? (await getDefaultLocale()) },
+                Layout,
+                Page,
+            };
+        },
+    } as Route<Match, Props | Match>;
+}
+
+type Route<Match, Props> = {
+    match(
+        path: string,
+        getDefaultLocale: () => Promise<string>,
+        isEnabledLocale: (locale: string) => Promise<boolean>,
+    ): Promise<
+        | {
+              match: Match;
+              Layout?: ServerComponentType<Props>;
+              Page: ServerComponentType<Props>;
+          }
+        | undefined
+    >;
+};
+
+export async function match(path: string | string[], routes: Route<any, any>[]) {
+    assertServerEnv('match');
+
+    const { contentDelivery } = api();
+
+    async function getDefaultLocale() {
+        return (await contentDelivery.language())!.locale.code;
```

Comment:
I think this should be typed better (e.g. when `localeCode` is not provided, the types should indicate that a value would always be returned).

Reply:
dc6e0c79c44eae36933d9b3ad9f6657d1e985cca